[Aaron Tan] iP

aarontangr opened the pull request:
## Duke
> “Well done, students—if you were trying to disappoint me.” – Master Shifu ([source](https://kungfupanda.fandom.com/wiki/Shifu/Quotes))

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
5. And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

aarontangr committed:
Level-1

aarontangr committed:
Level-2

aarontangr committed:
Level-4

aarontangr committed:
A-TextUiTesting

aarontangr committed:
Handle errors

aarontangr committed:
Added delete functionality

aarontangr committed:
Delete .class files

Commit:
Add read and save functionality

Commit:
Add date and time functionality

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

Commit:
Redo save functionality

Commit:
Merge branch 'branch-Level-7'

Commit:
Fix conflicts

Commit:
Fix date and time parsing

Commit:
Implement more OOP style

Commit:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

Commit:
Add javadocs to all non-private methods

Commit:
Add function to find a task by searching for a keyword.

Commit:
Update coding standard.

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

Commit:
Checkstyle

Commit:
Add basic GUI

Commit:
Merge branch 'branch-A-Checkstyle'

Commit:
Merge branch 'branch-Level-10'

gabyang reviewed (commented):
Wow what a great application! Overall had a great time reading it as it is very readable! A few minor changes can bring your application to the next level!

aarontangr commented:
> Wow what a great application!

Thanks!

gabyang commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,171 @@
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.Region;
+import javafx.scene.layout.VBox;
+import javafx.stage.Stage;
```
A very minor coding standard issue, can be easily solved by shifting it all the way down

gabyang commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+        return String.format("[D]%s (by: %s)",
+                super.toString(), date.format(formatter));
+    }
+}
```
Don't you think that separating this file along with Event, Task, Todo into another package called task would make it more organised? I think so!

gabyang commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Override
+    public void start(Stage stage) {
```
I do believe the code can be made even better if you were to extract out the UI code and put that in the Ui.java file! But that would just be my personal preference!

gabyang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,219 @@
+        this.tasks = tasks;
+        String[] words = s.split(" ");
+        String command = words[0];
+        switch (command) {
```
I like your use of switch statements! That makes your program more readable compared to if-else statements!

Rachel-AG replied:
Agreed!

gabyang commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,219 @@
+     *
+     * @param input String to be converted to a Todo.
+     */
+    protected void insertTodo(String input) {
```
Might I suggest that this method and others below it be abstracted out into their respective classes?

gabyang commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public abstract class Task {
+    protected String description;
+    protected boolean isDone;
```
Good choice of boolean variable name!

gabyang commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,64 @@
+package duke;
```
I may advise against the use of this class because it seems to me that this class is emulating the Arraylist class!

gabyang commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,17 @@
+package duke;
```
I definitely think that more system.out.println statements can be placed in here!

Rachel-AG commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+     *
+     * @return String in the format of D | isDone | description | date.
+     */
+    public String saveString() {
```
Perhaps you can consider changing the method name? Maybe toSaveFormat or toDatabaseString?

Rachel-AG commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,219 @@
+        try {
+            String toFind = s.substring(5);
+            ArrayList<Task> matchingTasks = new ArrayList<>();
+            for (int i = 0; i < tasks.size(); i++) {
+                Task task = tasks.get(i);
+                if (task.description.contains(toFind)) {
+                    matchingTasks.add(task);
+                }
+            }
```
Maybe you can consider moving this part to a new method in TaskList class that returns an instance of TaskList with matching tasks?

Rachel-AG commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,37 @@
+     * Test Deadline with valid input
+     */
+    @Test
+    public void test1() {
```
I think you can consider renaming the method to be more descriptive. Perhaps to whatIsBeingTested_descriptionOfTestInputs_expectedOutcome format?

Rachel-AG commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,219 @@
+     */
+    protected void insertTask(Task task) {
+        tasks.add(task);
+        System.out.println("added: ");
+        System.out.println("\t" + task);
+        System.out.format("you have %d task(s) in the list\n", tasks.size());
```
I think you can consider moving your System.out.println statements to Ui class.

Rachel-AG reviewed (commented):
Generally, I think the code is very readable and easy to follow. Great job :)

Commit:
Refactor parser to return String

Commit:
Add JavaFX

Commit:
Remove unused variables and imports

myangat0343 commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,39 @@
+ *
+ * @author Aaron Tan
+ */
+public class Todo extends Task {
```
Hii, I like your well-organized code. Maybe you could use packages, (e.g. Putting Todo.java, deadline.java into a new Task Package) for better organization : )

myangat0343 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,219 @@
+     * @param s String to be processed.
+     * @param tasks Task list to be operated on.
+     */
+    protected void process(String s, TaskList tasks) {
```
Using switch statements is well-organized! love you design!

myangat0343 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,219 @@
+     *
+     * @param input String that contains information about item to be deleted.
+     */
+    protected void deleteTask(String input) {
```
So many methods .... Maybe you could use a new class to handle these methods : ) it will be more readable : )

myangat0343 reviewed (commented):
Add some small comments : ) And your code is good : )

Commit:
Add assertions

Assertions are useful for verifying assumptions about internal
invariants and other various invariants.

Adding assertions are great to improve software and can show
where a programmer made a mistake.

aarontangr committed:
Merge pull request #1 from aarontangr/A-Assertions

Add assertions

Commit:
Improve code quality of package duke.

Making code understandable is important, improve it by trying to
apply more SLAP and KISS

aarontangr committed:
Merge branch 'master' into branch-A-CodeQuality

aarontangr committed:
Merge pull request #2 from aarontangr/branch-A-CodeQuality

Improve code quality of package duke.

Commit:
Implement CI/CD

Commit:
Update various files to follow style guide.

Commit:
Repackage files in duke

Commit:
Update imports to match checkstyle and remove redundant imports

Commit:
Update junit tests

Commit:
Update DialogBox to be able to fit long text

Commit:
Update text-ui-test to deprecate CLI redirection testing

Commit:
Add newline to end of EXPECTED.TXT

Commit:
Update gradle.yml

Commit:
Add reminder functionality

Commit:
Add user guide

Commit:
Fix .png file naming

Commit:
Remove datafile

Commit:
Update build.gradle to match Launcher

Commit:
Update various functions to throw DukeException

Commit:
Fix save file path

Commit:
Remove unused imports

Commit:
Deprecate IO redirection test

Commit:
Update javafx to v11.0.2

Commit:
Update storage to handle invalid save file

Commit:
Update font

Commit:
Add invalid save file handling

Commit:
Remove unused imports
